### PR TITLE
Rename Shipit::PullRequest to Shipit::MergeRequest

### PR DESCRIPTION
**Pull request by indiebrain**
Implements the proposal of Shopify/shipit-engine issue #1093 by
renaming the concept of the `Shipit::PullRequest` to
`Shipit::MergeRequest`. Consideration was given to maintaining the
name "PullRequest" where shipit-engine is interacting directly with
the Github API - for example, in the `Shipit.github.api client`. In
this way the `Shipit::MergeRequest` is related to Github Pull
Requests, but not necessarily _of_ a Github Pull Request.

The goal is to make way for a more direct modeling of a Github Pull
Request in shipit-engine - `Shipit::PullRequest` - which like
`Shipit::Commit` models the Github concept. The plan is to use this
closer modeling of the Github Pull Request to finish preparations for
submission of our "Review Stack" feature to `Shopify/shipit-engine`.

Notes
-------

This does include non-backward compatible HTTP API changes to 
shipit-engine - the `pull_requests` collection resources have been 
renamed to `merge_requests`. 

References
----------

- https://github.com/Shopify/shipit-engine/issues/1093

### CODE REVIEW

File: app/models/shipit/merge_request.rb [hunk 1]
```diff
@@ -112,17 +112,17 @@ def self.schedule_merges
     def self.extract_number(stack, number_or_url)
       case number_or_url
       when /\A#?(\d+)\z/
-        $1.to_i
+        Regexp.last_match(1).to_i
```
Comment (casperisfine) [inline app/models/shipit/merge_request.rb:115]: I just disabled that silly cop on master. Please revert that change.

Comment (indiebrain) [reply app/models/shipit/merge_request.rb]: [There was a another definition of that cop later on in the `.rubocop.yml`](https://github.com/Shopify/shipit-engine/pull/1095/commits/5dd9f8a1d03f1d2ec532c83ed3383dd1fa451789#diff-11a0d7906801d9dea0eccb85667ad811) which overrode the changes in https://github.com/Shopify/shipit-engine/commit/e0acb4e68805d04b119001a38c765c4c3358461b. I've removed the later cop definition so the `Style/PerlBackrefs` cop is totally disabled for the project as part of this body of work. Hope that doesn't muddy the waters too much.

Comment (casperisfine) [reply app/models/shipit/merge_request.rb]: Thanks.

File: app/jobs/shipit/merge_merge_requests_job.rb [hunk 2]
```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+module Shipit
+  class MergeMergeRequestsJob < BackgroundJob
```
Comment (casperisfine) [inline app/jobs/shipit/merge_merge_requests_job.rb:3]: I just noticed this. `MergeMerge` is a bit weird. I'm not sure what would be better though. Maybe `ProcessMergeRequestsJob`?

File: app/models/shipit/commit.rb [hunk 3]
```diff
@@ -219,7 +219,7 @@ def detach_children!
       children.detach!
     end
 
-    def pull_request?
+    def merge_request?
```
Comment (casperisfine) [inline app/models/shipit/commit.rb:222]: I think this shouldn't be changed. The meaning of this method is to tell wether the commit looks like a Pull Request merge commit.

File: app/models/shipit/commit.rb [hunk 4]
```diff
@@ -289,20 +289,20 @@ def deploy_failed?
 
     def identify_pull_request
```
Comment (casperisfine) [inline app/models/shipit/commit.rb:290]: However this should be renamed in `identify_merge_request `

File: app/jobs/shipit/merge_merge_requests_job.rb (hunk 2 above)
Comment (indiebrain) [reply app/jobs/shipit/merge_merge_requests_job.rb:3]: On a scale of 1-5, I'm a 5 (Strongly Agree). 😄

File: app/models/shipit/commit.rb (hunk 3 above)
Comment (indiebrain) [reply app/models/shipit/commit.rb:222]: Do you think the name `Commit#pull_request_merge?` would ever-so-slightly elevate that intent to the reader over `Commit#pull_request?`? Would `Commit#pull_request_merge?` be an acceptable name. Do you think it improves clarity at all for the reader? Or would you prefer we stick with `Commit#pull_request?` predicate?

Comment (casperisfine) [reply app/models/shipit/commit.rb:222]: I think it's slightly too verbose. I'd rather stick to `pull_request?` for now, we can always rename it later on.

File: app/jobs/shipit/merge_merge_requests_job.rb (hunk 2 above)
Comment (indiebrain) [reply app/jobs/shipit/merge_merge_requests_job.rb]: Resolved in: https://github.com/Shopify/shipit-engine/pull/1095/commits/ce8120a5f1e10b478c6b0facf5893a2027974d60

File: app/models/shipit/commit.rb (hunk 3 above)
Comment (indiebrain) [reply app/models/shipit/commit.rb]: Resolved in: https://github.com/Shopify/shipit-engine/pull/1095/commits/ce8120a5f1e10b478c6b0facf5893a2027974d60

File: app/models/shipit/commit.rb (hunk 4 above)
Comment (indiebrain) [reply app/models/shipit/commit.rb]: Resolved in: https://github.com/Shopify/shipit-engine/pull/1095/commits/ce8120a5f1e10b478c6b0facf5893a2027974d60

**Comment by benlangfeld** [comment]
@casperisfine Looks like this should be ready to go :) Thank you for your very agile reviews, we appreciate it.

**Comment by casperisfine** [comment]
@benlangfeld more than welcome o7

**Pull request merged**